### PR TITLE
[libc][math][c23] Add dadd{l,f128} and ddiv{l,f128} C23 math functions

### PR DESCRIPTION
**Pull request by aaryanshukla**
- fadd removed because I need to add for different input types
- finishing rest of basic operations
- noticed duplicates will remove

### CODE REVIEW

File: libc/spec/stdc.td
Comment (overmighty) [inline libc/spec/stdc.td:403]: `daddf128` and `ddivf128` should be LLVM libc extensions, so they should be moved to libc/spec/llvm_libc_ext.td. C23 only defines these:

```c
double daddl(long double x, long double y);
_FloatM fMaddfN(_FloatN x, _FloatN y); // M < N

double ddivl(long double x, long double y);
_FloatM fMdivfN(_FloatN x, _FloatN y); // M < N
```

Also, `daddf128` and `ddivf128` should be guarded by `LIBC_TYPES_HAS_FLOAT128`.

```suggestion
          FunctionSpec<"daddl", RetValSpec<DoubleType>, [ArgSpec<LongDoubleType>, ArgSpec<LongDoubleType>]>,

          FunctionSpec<"ddivl", RetValSpec<DoubleType>, [ArgSpec<LongDoubleType>, ArgSpec<LongDoubleType>]>,
```

File: libc/src/math/daddf128.h
Comment (overmighty) [inline libc/src/math/daddf128.h:2]: Nit: formatting.

```suggestion
//===-- Implementation header for daddf128 ----------------------*- C++ -*-===//
```

File: libc/src/math/daddf128.h
Comment (overmighty) [inline libc/src/math/daddf128.h:13]: We used to include libc/src/__support/macros/properties/types.h instead of including float128.h directly.

File: libc/src/math/daddl.h
Comment (overmighty) [inline libc/src/math/daddl.h:2]: Nit: formatting.

File: libc/src/math/ddivf128.h
Comment (overmighty) [inline libc/src/math/ddivf128.h:2]: Nit: formatting.

File: libc/src/math/ddivf128.h
Comment (overmighty) [inline libc/src/math/ddivf128.h:12]: We used to include libc/src/__support/macros/properties/types.h instead of including float128.h directly.

File: libc/src/math/ddivl.h
Comment (overmighty) [inline libc/src/math/ddivl.h:2]: Nit: formatting.

File: libc/src/math/fadd.h
Comment (overmighty) [inline libc/src/math/fadd.h:10]: Maybe `fadd` fixes should be moved to another, small PR.

File: libc/src/math/generic/CMakeLists.txt
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:547]: Maybe `fadd` fixes should be moved to another, small PR.

File: libc/src/math/generic/daddf128.cpp
Comment (overmighty) [inline libc/src/math/generic/daddf128.cpp:2]: Nit: formatting.

File: libc/src/math/generic/daddl.cpp
Comment (overmighty) [inline libc/src/math/generic/daddl.cpp:2]: Nit: formatting.

File: libc/src/math/generic/ddivf128.cpp
Comment (overmighty) [inline libc/src/math/generic/ddivf128.cpp:2]: Nit: formatting.

File: libc/src/math/generic/ddivl.cpp
Comment (overmighty) [inline libc/src/math/generic/ddivl.cpp:2]: Nit: formatting.

File: libc/test/src/math/daddf128_test.cpp
Comment (overmighty) [inline libc/test/src/math/daddf128_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for daddf128 --------------------------------------------===//
```

File: libc/test/src/math/daddl_test.cpp
Comment (overmighty) [inline libc/test/src/math/daddl_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for daddl -----------------------------------------------===//
```

File: libc/test/src/math/ddivl_test.cpp
Comment (overmighty) [inline libc/test/src/math/ddivl_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for ddivl -----------------------------------------------===//
```

File: libc/test/src/math/smoke/CMakeLists.txt
Comment (overmighty) [inline libc/test/src/math/smoke/CMakeLists.txt:4168]: The smoke tests themselves don't use `libc.src.__support.FPUtil.generic.add_sub`.

```suggestion
  DEPENDS
    libc.src.math.daddl
```

File: libc/test/src/math/smoke/daddf128_test.cpp
Comment (overmighty) [inline libc/test/src/math/smoke/daddf128_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for daddf128 --------------------------------------------===//
```

File: libc/test/src/math/smoke/daddl_test.cpp
Comment (overmighty) [inline libc/test/src/math/smoke/daddl_test.cpp:1]: Nit: formatting.

```suggestion
//===-- Unittests for daddl -----------------------------------------------===//
```

File: libc/test/src/math/CMakeLists.txt
Comment (overmighty) [inline libc/test/src/math/CMakeLists.txt:59]: The unit tests themselves don't use `libc.src.__support.FPUtil.generic.add_sub`.

```suggestion
  DEPENDS
    libc.src.math.daddl
```

**Comment by overmighty** [comment]
I had made a Python script to generate some of the boilerplate for math functions, but it's not yet updated for the move to `LIBC_NAMESPACE_DECL`. It should save you time and avoid all the formatting nits.

https://github.com/overmighty/llvm-dev-utils/blob/master/llvm-libc-add-math-function.py

File: libc/config/gpu/entrypoints.txt
Comment (overmighty) [reply libc/config/gpu/entrypoints.txt:260]: `long double` shouldn't be used on GPUs either, actually. AMDGPU will use a different format than the host, and apparently NVIDIA's proprietary compiler will emit a warning if you use `long double` in device code. So none of these entrypoints should be enabled on GPUs.

File: libc/spec/stdc.td
Comment (overmighty) [reply libc/spec/stdc.td:403]: Also, functions that are LLVM libc extensions should have a `|check|\*` in the implementation status table instead of just a `|check|`.

File: libc/newhdrgen/yaml_to_classes.py
Comment (aaryanshukla) [reply libc/newhdrgen/yaml_to_classes.py:121]: was working on it before in another branch, so it got it from there. During merge conflicts I'll refer to the new addition,

File: libc/config/baremetal/arm/entrypoints.txt
Comment (aaryanshukla) [reply libc/config/baremetal/arm/entrypoints.txt]: addressed all your comments in the most recent commit

Comment (overmighty) [reply libc/config/baremetal/arm/entrypoints.txt]: https://github.com/llvm/llvm-project/pull/100456/files shows unresolved comments.

File: libc/config/darwin/x86_64/entrypoints.txt
Comment (overmighty) [inline libc/config/darwin/x86_64/entrypoints.txt:125]: There were previously no `float128` entrypoints in this file either so there might be an issue with `float128` on this target too.

File: libc/config/baremetal/arm/entrypoints.txt
Comment (aaryanshukla) [reply libc/config/baremetal/arm/entrypoints.txt]: just resolved

File: libc/src/math/daddl.h
Comment (overmighty) [inline libc/src/math/daddl.h:1]: Nit: formatting.

```suggestion
//===-- Implementation header for daddl ------------------------*- C++ -*-===//
```

File: libc/src/math/generic/daddf128.cpp
Comment (overmighty) [inline libc/src/math/generic/daddf128.cpp:10]: Existing code transitively includes .../properties/types.h from src/math/<function>.h, so I would do the same.

```suggestion
```

File: libc/src/math/generic/ddivf128.cpp
Comment (overmighty) [inline libc/src/math/generic/ddivf128.cpp:10]: Existing code transitively includes .../properties/types.h from src/math/<function>.h, so I would do the same.

```suggestion
```

File: libc/src/math/generic/CMakeLists.txt
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:142]: `daddl` doesn't use `float128`.

```suggestion
  DEPENDS
    libc.src.__support.FPUtil.generic.add_sub
```

File: libc/src/math/generic/CMakeLists.txt
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:169]: `ddivl` doesn't use `float128`.

```suggestion
  DEPENDS
    libc.src.__support.FPUtil.generic.div
```

File: libc/src/math/generic/CMakeLists.txt
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:143]: Very tiny nit: blank line.

```suggestion
```

File: libc/config/linux/arm/entrypoints.txt
Comment (overmighty) [reply libc/config/linux/arm/entrypoints.txt:233]: This wasn't resolved.

File: libc/config/linux/riscv/entrypoints.txt
Comment (overmighty) [reply libc/config/linux/riscv/entrypoints.txt:387]: This wasn't resolved.

File: libc/test/src/math/CMakeLists.txt
Comment (overmighty) [reply libc/test/src/math/CMakeLists.txt:59]: This wasn't resolved.

File: libc/config/baremetal/arm/entrypoints.txt
Comment (lntue) [inline libc/config/baremetal/arm/entrypoints.txt:250]: Don't enable entrypoints for baremetal yet, since we cannot test them downstream, and we don't want to revert the whole PR if downstream baremetal bots fail.

File: libc/config/linux/x86_64/entrypoints.txt
Comment (lntue) [inline libc/config/linux/x86_64/entrypoints.txt:387]: `daddf128` and `ddivf128` should be under the `LIBC_TYPES_HAS_FLOAT128` guard below.

File: libc/newhdrgen/yaml_to_classes.py
Comment (lntue) [inline libc/newhdrgen/yaml_to_classes.py:176]: sync your PR to head to resolve the merge conflict.

**Comment by overmighty** [comment]
I've changed the PR title to better reflect the fact that it adds multiple variants of `dadd*` and `ddiv*`, if you don't mind.

Be careful of what you set the PR description to, as GitHub will suggest that as extended commit description by default when merging.

**Review by lntue — APPROVED** [review]

File: libc/test/src/math/CMakeLists.txt
Comment (overmighty) [inline libc/test/src/math/CMakeLists.txt:100]: None of the tests directly depend on the FPUtil.generic.* stuff. Not just `daddl_test`.

```suggestion
  DEPENDS
    libc.src.math.daddf128
)

add_fp_unittest(
  ddivf128_test
  NEED_MPFR
  SUITE
    libc-math-unittests
  SRCS
    ddivf128_test.cpp
  HDRS
    DivTest.h
  DEPENDS
    libc.src.math.ddivf128
)

add_fp_unittest(
  ddivf128_test
  NEED_MPFR
  SUITE
    libc-math-unittests
  SRCS
    ddivf128_test.cpp
  HDRS
    DivTest.h
  DEPENDS
    libc.src.math.ddivf128
```

File: libc/newhdrgen/yaml_to_classes.py
Comment (overmighty) [inline libc/newhdrgen/yaml_to_classes.py:177]: You can just remove this new line so that this PR doesn't change newhdrgen at all.

```suggestion
```

File: libc/docs/math/index.rst
Comment (overmighty) [inline libc/docs/math/index.rst:139]: Kind of nitpicky, but technically I think this should be in its own PR, since:

> Once you have a patch ready, it is time to submit it. The patch should:
> - [...]
> - not contain any unrelated changes
> - be an isolated change. Independent changes should be submitted as separate patches as this makes reviewing easier.

https://llvm.org/docs/Contributing.html#how-to-submit-a-patch

I don't know if this gets a pass for being such a small change or not. @lntue

Comment (aaryanshukla) [reply libc/docs/math/index.rst:139]: will fix this

Comment (lntue) [reply libc/docs/math/index.rst]: I think it's ok for this small docs update as it's tangent to this PR, just like sometime we fix some thing right next to what we update.

Comment (overmighty) [reply libc/docs/math/index.rst]: Right.

File: libc/test/src/math/CMakeLists.txt
Comment (overmighty) [inline libc/test/src/math/CMakeLists.txt:98]:
```suggestion
  DEPENDS
    libc.src.math.ddivf128
```

**Review by overmighty — APPROVED** [review]
Just this last request and it LGTM. You can add the `fadd` docs change back if you want, since Tue approves of it. I forgot that we sometimes also change some `COMPILE_OPTIONS -O2` to `-O3` when changing/adding functions nearby.

File: libc/test/src/math/CMakeLists.txt
Comment (aaryanshukla) [reply libc/test/src/math/CMakeLists.txt]: just made last minute changes if you want to check those out

Comment (overmighty) [reply libc/test/src/math/CMakeLists.txt]: Thanks, I didn't catch that.

**Review by overmighty — APPROVED** [review]

File: libc/test/src/math/daddf128_test.cpp
Comment (overmighty) [inline libc/test/src/math/daddf128_test.cpp:1]: I missed the fact that these were MPFR tests using `float128`. `float128` isn't supported for MPFR tests as I mentioned on your other PR.

Comment (aaryanshukla) [reply libc/test/src/math/daddf128_test.cpp:1]: fixed, let me if its all good I think it'll fix the build breaking as well.

File: libc/test/src/math/ddivl_test.cpp
Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: fixed + it passes for me.

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: Are you on an x86-64 or AArch64 machine?

**Comment by aaryanshukla** [comment]
X86-64

File: libc/test/src/math/ddivl_test.cpp
Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: X86-64

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: Interesting. I'll wait for the CI results.

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: Build broke. Will check it out.

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: Any thoughts as to why? I just changed the AddTest to DivTest for Hdrs in the tests cmakelists

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: When math tests fail only for `long double`, it's usually because the format of `long double` on x86 needs special care. I would suggest you remove `ddivl` from this PR so that we can merge it. Then, if you want to fix `ddivl` yourself, you can read https://en.wikipedia.org/wiki/Extended_precision#x86_extended_precision_format.

I'm not sure why the tests passed on your machine. Could you enter these commands in a terminal and show the output?

```sh
builddir=replace/with/path/to/your/build/directory
cmake --build "${builddir}" --target libc.test.src.math.ddivl_test.__unit__.__build__
file "${builddir}"/projects/libc/test/src/math/libc.test.src.math.ddivl_test.__unit__.__build__
```

File: libc/docs/math/index.rst
Comment (overmighty) [inline libc/docs/math/index.rst:119]:
```suggestion
| ddiv             | N/A              | N/A             |                        | N/A                  | |check|\*              | 7.12.14.4              | F.10.11                    |
```

**Review by overmighty — COMMENTED** [review]
I missed that the docs weren't updated after the removal of `ddivl`.

**Review by overmighty — APPROVED** [review]

File: libc/test/src/math/ddivl_test.cpp
Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: @aaryanshukla Could you please run these commands and post the output?

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: ninja: error: unknown target 'libc.test.src.math.ddivl_test.__unit__.__build__', did you mean 'libc.test.src.math.fdiml_test.__unit__.__build__'?
build/projects/libc/test/src/math/libc.test.src.math.ddivl_test.__unit__.__build__: cannot open `build/projects/libc/test/src/math/libc.test.src.math.ddivl_test.__unit__.__build__' (No such file or directory)

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: I deleted and created build again, still same issue.

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: When the CI failed on another PR before you added explicit template instantiations in MPFRUtils.cpp, did you not get that error while testing locally? Maybe the MPFR tests are just disabled in your build. They're disabled in full build mode.

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: When turning full build off, for NO_FMA_OPT DDIVL I am getting all cases passed.

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: file build/projects/libc/test/src/math/libc.test.src.math.ddivl_test.__unit__.__NO_FMA_OPT.__build__
build/projects/libc/test/src/math/libc.test.src.math.ddivl_test.__unit__.__NO_FMA_OPT.__build__: cannot open `build/projects/libc/test/src/math/libc.test.src.math.ddivl_test.__unit__.__NO_FMA_OPT.__build__' (No such file or directory)

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: This is my ninja check-libc tests: Ran 5 tests.  PASS: 5  FAIL: 0
[795/1116] Running unit test libc.test.src.math.smoke.ddivl.__unit__.__NO_FMA_OPT
[==========] Running 5 tests from 1 test suite.
[ RUN      ] LlvmLibcDivTest.SpecialNumbers
[       OK ] LlvmLibcDivTest.SpecialNumbers (5 us)
[ RUN      ] LlvmLibcDivTest.DivisionByZero
[       OK ] LlvmLibcDivTest.DivisionByZero (3 us)
[ RUN      ] LlvmLibcDivTest.InvalidOperations
[       OK ] LlvmLibcDivTest.InvalidOperations (5 us)
[ RUN      ] LlvmLibcDivTest.RangeErrors
[       OK ] LlvmLibcDivTest.RangeErrors (12 us)
[ RUN      ] LlvmLibcDivTest.InexactResults
[       OK ] LlvmLibcDivTest.InexactResults (1 us)

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: You should see a warning message saying "Math tests using MPFR will be skipped" in the CMake configuration output by the way. You can apply this patch and check the CMake configuration output to try to debug the issue:

```diff
diff --git a/libc/cmake/modules/LLVMLibCCheckMPFR.cmake b/libc/cmake/modules/LLVMLibCCheckMPFR.cmake
index a27c2dc0c030..3672d25d1378 100644
--- a/libc/cmake/modules/LLVMLibCCheckMPFR.cmake
+++ b/libc/cmake/modules/LLVMLibCCheckMPFR.cmake
@@ -14,5 +14,7 @@ else()
     ${LIBC_SOURCE_DIR}/utils/MPFRWrapper/check_mpfr.cpp
     LINK_LIBRARIES
       -lmpfr -lgmp -latomic
+    OUTPUT_VARIABLE foo
   )
+  message(WARNING ${foo})
 endif()
```

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: Yeah I see the math tests using MPFR will be skipped, I need to install MPFR. Should I still merge this patch then try to solve the MPFR issues.

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: The CI build passed after you removed `ddivl` so you can merge this PR.

Comment (aaryanshukla) [reply libc/test/src/math/ddivl_test.cpp]: Is there documentation for installing mpfr.

Comment (overmighty) [reply libc/test/src/math/ddivl_test.cpp]: I'm not aware of any. If you're on Linux then you should search your Linux distribution's package repositories.

**Pull request merged**
